File: backend/src/main/java/com/mgs/ergomanager/service/UserService.java
package com.mgs.ergomanager.service;

import com.mgs.ergomanager.dto.user.UserRequestDTO;
import com.mgs.ergomanager.dto.user.UserResponseDTO;
import com.mgs.ergomanager.dto.user.UserUpdateRequestDTO;

import java.util.List;

/**
 * Management of the administrators and the ergonomists.
 */
public interface UserService {

    /**
     * Returns every registered user.
     *
     * @return list of users
     */
    List<UserResponseDTO> findAll();

    /**
     * Returns a single user.
     *
     * @param id identifier of the user
     * @return the user
     */
    UserResponseDTO findById(Long id);

    /**
     * Registers a new user and hashes the received password.
     *
     * @param request data of the user
     * @return the created user
     */
    UserResponseDTO create(UserRequestDTO request);

    /**
     * Updates the data of an existing user.
     *
     * @param id      identifier of the user
     * @param request new data of the user
     * @return the updated user
     */
    UserResponseDTO update(Long id, UserUpdateRequestDTO request);

    /**
     * Deactivates a user so it can no longer sign in.
     *
     * @param id identifier of the user
     */
    void deactivate(Long id);
}

File: frontend/src/app/pages/users/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { UserRequest, UserResponse } from '../../shared/models/user.model';

/**
 * HTTP access to the /api/users endpoints.
 */
@Injectable({ providedIn: 'root' })
export class UserService {
    private readonly http = inject(HttpClient);

    private readonly baseUrl = `${environment.apiUrl}/users`;

    /**
     * Reads every registered user.
     *
     * @returns list of users
     */
    findAll(): Observable<UserResponse[]> {
        return this.http.get<UserResponse[]>(this.baseUrl);
    }

    /**
     * Reads a single user.
     *
     * @param id identifier of the user
     * @returns the user
     */
    findById(id: number): Observable<UserResponse> {
        return this.http.get<UserResponse>(`${this.baseUrl}/${id}`);
    }

    /**
     * Registers a new administrator or ergonomist.
     *
     * @param request data typed in the form
     * @returns the created user
     */
    create(request: UserRequest): Observable<UserResponse> {
        return this.http.post<UserResponse>(this.baseUrl, request);
    }

    /**
     * Updates an existing user.
     *
     * @param id      identifier of the user
     * @param request data typed in the form
     * @returns the updated user
     */
    update(id: number, request: UserRequest): Observable<UserResponse> {
        return this.http.put<UserResponse>(`${this.baseUrl}/${id}`, request);
    }

    /**
     * Deactivates a user so it can no longer sign in.
     *
     * @param id identifier of the user
     * @returns an empty response
     */
    deactivate(id: number): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/${id}`);
    }
}
